domain/src/test/java/org/oppia/domain/oppialogger/exceptions/ExceptionsControllerTest.kt

```diff
@@ -0,0 +1,350 @@
+    assertThat(exception.cause?.message).isEqualTo(exceptionThrown.cause?.message)
+    assertThat(exception.cause?.stackTrace).isEqualTo(exceptionThrown.cause?.stackTrace)
+    assertThat(exceptionLog.exceptionType).isEqualTo(ExceptionType.NON_FATAL)
+  }
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testController_logException_Fatal_withNoNetwork_logsToCacheStore() {
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+    val exceptionThrown = Exception("TEST MESSAGE", Throwable("TEST"))
+    exceptionsController.logFatalException(exceptionThrown, TEST_TIMESTAMP)
+
+    val cachedExceptions = exceptionsController.getExceptionLogs()
+    cachedExceptions.observeForever(mockOppiaExceptionLogsObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockOppiaExceptionLogsObserver, atLeastOnce())
+      .onChanged(oppiaExceptionLogsResultCaptor.capture())
+
+    val exceptionLog = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(0)
+    val exception = exceptionLogToException(exceptionLog)
+    assertThat(exception.message).isEqualTo(exceptionThrown.message)
+    assertThat(exception.stackTrace).isEqualTo(exceptionThrown.stackTrace)
+    assertThat(exception.cause?.message).isEqualTo(exceptionThrown.cause?.message)
+    assertThat(exception.cause?.stackTrace).isEqualTo(exceptionThrown.cause?.stackTrace)
+    assertThat(exceptionLog.exceptionType).isEqualTo(ExceptionType.FATAL)
+  }
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testController_logExceptions_exceedLimit_checkCorrectEviction() {
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+
+    exceptionsController.logFatalException(Exception("TEST1", Throwable("ONE")), 1556094120000)
+    exceptionsController.logNonFatalException(Exception("TEST2", Throwable("TWO")), 1556094110000)
+    exceptionsController.logFatalException(Exception("TEST3", Throwable("THREE")), 1556094100000)
+    exceptionsController.logFatalException(Exception("TEST4", Throwable("FOUR")), 1556094000000)
+
+    val cachedExceptions = exceptionsController.getExceptionLogs()
+    cachedExceptions.observeForever(mockOppiaExceptionLogsObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockOppiaExceptionLogsObserver, atLeastOnce())
+      .onChanged(oppiaExceptionLogsResultCaptor.capture())
+
+    val exceptionOne = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(0)
+    val exceptionTwo = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(1)
+    val cacheStoreSize = oppiaExceptionLogsResultCaptor.value.getOrThrow().exceptionLogList.size
+
+    assertThat(cacheStoreSize).isEqualTo(2)
+    // In this case, 3 FATAL and 1 NON-FATAL exception was logged. So while pruning, none of the retained logs should have NON-FATAL exception type.
```

Comment:
Also regarding the actual eviction strategy, shouldn't we prune non-fatal logs in favor of keeping fatal ones? Those seem more important to log to me.

Reply:
Yes, that is what's happening. We prune non-fatal exceptions in favour of the fatal ones.

Reply:
The test suggests otherwise though. Isn’t it showing that an older fatal exception is being pruned in favor of a newer non fatal exception?

Reply:
No, the order of logging is fatal->non-fatal->fatal->fatal. So it prunes the non-fatal exception when it adds the third log (which is fatal). I've updated the comment to make it clearer.

Reply:
Sorry, I think I’m confused. The test is verifying that one of the pending exceptions is non-fatal—how does that correspond to the exception being pruned? How exactly is this test verifying the pruning strategy?

Reply:
The test is verifying that the exception type of any of the stored exception logs is not equal to non-fatal. It verifies the pruning strategy as - 
- Non-fatal exceptions are pruned in favour of the fatal ones. 
- The least recent exception gets pruned in favour of the most recent one.

Reply:
The test is verifying that both the exceptions that were retained are not non-fatal (see the isNotEqualTo checks) . If I understand correctly, there were 4 total exceptions - 3 fatal, and one non-fatal.
What this test is doing is verifying that the non-fatal exception was pruned and the least recent fatal exception was pruned. Correct me if I'm wrong @Sarthak2601 ;

Reply:
Ah, I see where my confusion was. I actually kept reading ``isEqualTo``--thanks for the additional clarification!